src/main/java/org/yomigae/model/TempleModel.java

```diff
@@ -30,142 +30,186 @@
 
-		public static ToriiType fromIndex(int i) {
-			return values[i];
+		for (int i = 0; i < 5; ++i) {
+			submodels.add(createTorii(direction, ToriiModel.ToriiType.T1, t, tunnelKeys));
+			t.translate(i < 5 - 1 ? 1 + 11 / 12.f : 0, 0, 0);
 		}
 	}
 
-	public static class ToriiConfig {
+	private static ToriiModel createTorii(int direction, ToriiModel.ToriiType type, LXTransform t, List<String> toriiKeys) {
+		ToriiModel torii = new ToriiModel(type, direction, t, toriiKeys);
+		t.translate(torii.columnDepth * direction, 0, 0);
+		return torii;
 	}
 
-	public static class Torii extends LXModel {
-		private TempleModel.ToriiType type;
-		private float centerOffset;
-		private float openingWidth;
-		private float columnWidth, columnHeight, columnDepth;
-		private float beamHeight, beamLength;
-		private float eaveDepth; // calculated
+	private static LXModel[] buildSubmodels(LXTransform t) {
+		List<LXModel> submodels = new ArrayList<>();
 
-		private List<DirectionalPoint> frontPoints;
-		private List<DirectionalPoint> rearPoints;
+		t.push();
+		buildHalfTemple(submodels, -1, t);
+		t.pop();
 
-		public Torii(ToriiType type, float centerOffset, float openingWidth,
-				float columnWidth, float columnHeight, float columnDepth,
-				float beamHeight, float beamLength,
-				List<DirectionalPoint> frontPoints, List<DirectionalPoint> rearPoints) {
+		t.push();
+		buildHalfTemple(submodels, 1, t);
+		t.pop();
 
-			super(Lists.newArrayList(Iterables.concat(frontPoints, rearPoints)));
 
-			this.type = type;
-			this.centerOffset = centerOffset;
-			this.openingWidth = openingWidth;
-			this.columnWidth = columnWidth;
-			this.columnHeight = columnHeight;
-			this.columnDepth = columnDepth;
-			this.beamHeight = beamHeight;
-			this.beamLength = beamLength;
+		return submodels.toArray(new LXModel[0]);
+	}
 
-			eaveDepth = (beamLength - 2 * columnWidth - openingWidth) / 2;
+	public Set<LXPoint> filterPoints(Set<FilterFlags> flags) {
+		Set<LXPoint> filteredPoints = new HashSet<>(getPoints());
 
-			this.frontPoints = frontPoints;
-			this.rearPoints = rearPoints;
+		if (!flags.contains(FilterFlags.TWELVE)) {
+			filteredPoints.removeAll(twelveOclockPoints);
 		}
-
-		public TempleModel.ToriiType getType() {
-			return type;
+		if (!flags.contains(FilterFlags.SIX)) {
+			filteredPoints.removeAll(sixOclockPoints);
 		}
-
-		public List<DirectionalPoint> getNorthPoints() {
-			return rearPoints;
+		if (!flags.contains(FilterFlags.NINE)) {
+			filteredPoints.removeAll(nineOclockPoints);
+		}
+		if (!flags.contains(FilterFlags.THREE)) {
+			filteredPoints.removeAll(threeOclockPoints);
+		}
+		if (!flags.contains(FilterFlags.TUNNEL)) {
+			filteredPoints.removeAll(tunnelPoints);
+		}
+		if (!flags.contains(FilterFlags.HALL)) {
+			filteredPoints.removeAll(hallPoints);
 		}
 
-		public List<DirectionalPoint> getSouthPoints() {
-			return frontPoints;
+		return filteredPoints;
+	}
+
+	public Set<LXPoint> filterClusterPoints(Set<FilterFlags> flags) {
```

Comment:
When do you think we'll need to be able to filter out clusters, rather than individual points?